### DISCUSSION TITLE
"An account with the same identifier exists already" - But what's next?

### DISCUSSION THREAD

**Discussion by marcinkoziej** [ory/kratos#1042] [Q&A] [answered]
Hi!
Successfully implemented Kratos browser self service flows. Thanks for awesome work Ory!

I have a question about a specific flow:
1. I have enabled `password` and `oidc` methods
2. I have registered an account for email + password with example@gmail.com email. I can sign in with user/password to this identity.
3. When I try to sign in using OIDC with google, and select example@gmail.com, I receive and error and rightfully so: "An account with the same identifier (email, phone, username, ...) exists already.". 

However, this error is shown on a Register page, and the OIDC register form has displayed input fields for identity traits (first name, last name, email). What does Kratos want me to do now? I cannot obviously register again using already used email. What would be the sense to provide other traits? 
Shouldn't the right behaviour be to redirect me to login flow, but ask to log in via password method (and then maybe link the OIDC account?)

```
> kratos version
Version:			v0.5.4-alpha.1
Build Commit:	b02926c42aee2748bc37ce2600596bd0c2537a0d
Build Timestamp:	2020-11-11T13:00:29Z
```
 
The payload returned on register page using "get browser register flow" using the flowId provided by Kratos is pasted below.
You can see that the register form fields have values set to my first name, last name, email fetched from google. But for what reason? I can submit the form just to get the same error again,
```
{
  "id": "9bc83242-d38c-4910-ae5c-0808c2a0422f",
  "type": "browser",
  "expires_at": "2021-02-07T10:58:35.63459Z",
  "issued_at": "2021-02-07T09:58:35.63459Z",
  "request_url": "http://proca.lvh.me/self-service/methods/oidc/callback/google?state=2fe2d2b6-a3f6-4d64-a661-ebb56fb636ab&code=4%2F0AY0e-g74_3GkK17B5kF3RsDq9jR--_Zls4uh_8DIxH1a4CtGYY23ZDUwpbVY8ApQ2E3wkQ&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid&authuser=0&prompt=none",
  "active": "oidc",
  "messages": null,
  "methods": {
    "oidc": {
      "method": "oidc",
      "config": {
        "action": "https://proca.lvh.me/.ory/kratos/public/self-service/methods/oidc/auth/9bc83242-d38c-4910-ae5c-0808c2a0422f",
        "method": "POST",
        "fields": [
          {
            "name": "csrf_token",
            "type": "hidden",
            "required": true,
            "value": "WAn2K8llgxiVrCoSK5y4ts/0hSoRDezUrwQYUsLEmYGZLYZStDITmef0e8aSyCnq/btuGwGCnLVj2Ynr89Q4KQ=="
          },
          {
            "name": "traits.email",
            "type": "text",
            "value": "marcin@foobar.pl"
          },
          {
            "name": "traits.first_name",
            "type": "text",
            "value": "Marcin"
          },
          {
            "name": "traits.last_name",
            "type": "text",
            "value": "Ossowski"
          },
          {
            "name": "provider",
            "type": "submit",
            "value": "google"
          }
        ],
        "messages": [
          {
            "id": 4000007,
            "text": "An account with the same identifier (email, phone, username, ...) exists already.",
            "type": "error",
            "context": {}
          }
        ]
      }
    },
    "password": {
      "method": "password",
      "config": {
        "action": "https://proca.lvh.me/.ory/kratos/public/self-service/registration/methods/password?flow=9bc83242-d38c-4910-ae5c-0808c2a0422f",
        "method": "POST",
        "fields": [
          {
            "name": "csrf_token",
            "type": "hidden",
            "required": true,
            "value": "WAn2K8llgxiVrCoSK5y4ts/0hSoRDezUrwQYUsLEmYGZLYZStDITmef0e8aSyCnq/btuGwGCnLVj2Ynr89Q4KQ=="
          },
          {
            "name": "password",
            "type": "password",
            "required": true
          },
          {
            "name": "traits.email",
            "type": "email"
          },
          {
            "name": "traits.first_name",
            "type": "text"
          },
          {
            "name": "traits.last_name",
            "type": "text"
          }
        ]
      }
    }
  }
}
```

**Comment by vinckr** [comment]
> Shouldn't the right behaviour be to redirect me to login flow, but ask to log in via password method (and then maybe link the OIDC account?)
Makes sense yes.

Wouldnt `An account with the same identifier (email, phone, username, ...) exists already.` mean that account enumeration attacks are possible?
There is an open [issue](https://github.com/ory/kratos/issues/133) for that, but I think it should already have been solved for registration. 

Someone else had a similar problem: https://github.com/ory/kratos/issues/605#issuecomment-723111671 . Maybe that can help you out a bit.

**Reply by vinckr** [reply]
Hey @marcinkoziej 
Just checking in if your question was answered or if there was still something unclear. 
If not, could you select the right answer please? 

Thanks a ton!

**Comment by Serhii-the-Dev** [comment]
I have the same issue, if `traits.email` corresponds to an email returned by OIDC provider, Kratos returns `4000007` on both login and registration flows. Do we have an issue that covers this exact case?

**Comment by aeneasr** [comment, marked as answer]
This means that the email is already signed up. If it is, you need to link your account instead: https://www.ory.sh/kratos/docs/self-service/flows/user-settings#link-and-unlink-from-google-facebook-github--openid-connect--oauth-20

**Reply by Serhii-the-Dev** [reply]
Thank you for the explanation, running the `link` action for a particular identity should cover this case.

**Reply by EiffelFly** [reply]
@aeneasr Thanks for the explanation! 

But consider this scenario: If we only accept Oauth login, user A sign-in with client-A, later user A forgot he had already sign in with client A and try to sign-in with client-B, user A will face `An account with the same identifier (email, phone, username, ...) exists already.` issue.

As you mentioned above we can use kratos user-setting flow to solve this issue, but at that moment, user haven't login, we don't have kratos session, we can't verify setting flow to link user A's account to client-b.

What do you suggest to solve this kind of issue?

Thanks you again!!

**Reply by aeneasr** [reply]
We wanted to add a lookup to help the user, e.g. „you have signed up with google, please sign in first to link facebook.“

But this is not yet implemented :/

**Reply by EiffelFly** [reply]
HI, thanks for the quick clarification, do you suggest any workaround here?

**Reply by aeneasr** [reply]
Unfortunately, but fixing this behavior in Kratos shouldn't be too hard if you want to give it a try!

**Reply by EiffelFly** [reply]
@aeneasr, thanks for the clarification, but you mean directly modify the data in Kratos database or thought kratos setting flow?

**Reply by aeneasr** [reply]
Add the feature in Ory Kratos code

**Reply by EiffelFly** [reply]
Sure!! I am willing to help, could you shed light on where to start?

**Reply by aeneasr** [reply]
I think we need to check the error around here:

https://github.com/ory/kratos/blob/648a7bb36ae80fee1eb8279bd1c022c322edb996/selfservice/strategy/oidc/strategy_registration.go#L259

The error is coming from here:

https://github.com/ory/kratos/blob/77677f6b688202233d18ecbe96a9336943212489/selfservice/flow/registration/hook.go#L121-L126

So we should probably check for that error there and then return a different error message. This would also definitely need some tests!

**Reply by EiffelFly** [reply]
So what you suggest is show different message, then activate some interior procedure to make the user-agent know what they can do next? But I believe this interior process is not available too?

**Reply by aeneasr** [reply]
There's no interior process needed, just show a nice UI message to help the user understand what they need to do :)

**Reply by EiffelFly** [reply]
Got it!! I will try to implement it. Thanks for shedding the light
